### ISSUE TITLE
Error: Attempted to assign to readonly property.

### ISSUE THREAD

**Issue by gusga88** [bpampuch/pdfmake#1005]
Hello,
I am using pdfmake with ionic.

When I try to create a pdf like this (StateService.createPDF contains the DocDefinition.):
```js
angular.module('starter.services')
    .factory('CreatePdfService', function ($q,StateService) {
        var generateLayout = function(){
            var layout = StateService.createPDF;
            return layout;
        }

        var buildPDF = function() {
            var layout = generateLayout();
            return pdfMake.createPdf(layout);
        }
        return{
            generateLayout:generateLayout,
            buildPDF:buildPDF
        }
    });
```

I get the following error:

```
[Error] Error: Attempted to assign to readonly property.
createPdfKitDocument@file:///...../www/lib/pdfmake/build/pdfmake.js:2371:16
```

As soon as i use it like this I don't get any Errors:
```js
 angular.module('starter.services')
    .factory('CreatePdfService', function ($q) {
        var generateLayout = function(){
            return{
                content: [
                    'First paragraph',
                    'Another paragraph, this time a little bit longer to make sure, this line will be divided into at least two lines'
                ]
            };
        }

        var buildPDF = function(docDefinition) {
            var layout = generateLayout();
            return pdfMake.createPdf(docDefinition);
        }

        var base64ToUnit8Array = function(base64){
            var raw = atob(base64);
            var uint8Array = new Uint8Array(raw.length);
            for (var i = 0; i < raw.length; i++) {
                uint8Array[i] = raw.charCodeAt(i);
            }
            return uint8Array;
        }

        return{
            generateLayout:generateLayout,
            buildPDF:buildPDF
        }
    });
```

I need to get the docDefinition from my server. So I can't simply write it out in my code.

**Comment by liborm85** [comment]
In `pdfmake.js` on line 2371 is this code:
```js
docDefinition.images = docDefinition.images || {};
```
Pdfmake adds to the passed docDefinition various needed variables.

I do not know what `StateService.createPDF` doing, but it is probably readonly object. Try to clone this object and then insert it into pdfmake.

**Comment by gusga88** [comment]
Man that seems like such a rookie mistake. Thanks a lot for the quick answer.

**Comment by gusga88** [comment]
So I tried cloning the object but I still get the error. `StateService.createPDF` contains the DocDefinition.
StateService looks like this:
`angular.module('starter.services')
.value('StateService',{
    loggedin:true,
    showWorktimes:"",
    multiInputs:"",
    autoWorkState:"",
    receiverUrl:"",
    createPositions:"-1",
    createPDF:"",
    features:[]
});`

**Comment by liborm85** [comment]
I don't know what angular with docDefinition doing... Try to ask the angular forum, how to get a "normal" object from it.
Maybe this hack might work: `JSON.parse(JSON.stringify( StateService.createPDF ))`

**Comment by gusga88** [comment]
I found the solution:
`angular.fromJson(StateService.createPDF)`
The problem was, as you also presumed. That createPDF was just a JSON string and not an object.
Thanks again for the help.

**Comment by Saiheng** [comment]
I had the same issue on react-native. I also tried the `JSON.parse(JSON.stringify( docDefinition ))` approach as well as using `Ramda.clone(docDefinition)`. Neither solution did work. But I found that for me (strange enough) it was "hardcoded" styledefinitions which where causing the issues, e.g. things like:
`{text: 'Text', fontSize: 14, bold: false}`
If I removed them like this:
`{text: 'Text'}`
Things were working fine. 
So what I ended up doing was creating a big stylesheet for all my definitions and declared them in the definitions `styles`-property. To prevent these styles from causing the same error, I found that setting them through an anonymous function did the trick: 
`  styles: function () {
        return {
          header: {
            fontSize: 14,
            bold: true
          }
}`
Hope I could help someone.
